third_party/move/move-model/src/builder/model_builder.rs

```diff
@@ -460,6 +457,161 @@ impl<'env> ModelBuilder<'env> {
+                parent_name, field_name, child_name
+            ));
+        }
+        loop_notes.push(format!(
+            "`{}` contains field `{}: {}`, which forms a loop.",
+            self.get_struct_type_name(&path.last().unwrap().2, this_struct_id),
+            self.env.symbol_pool().string(path.last().unwrap().1),
+            self.get_struct_display_name_simple(this_struct_id)
+        ));
+        loop_notes
+    }
+
+    /// Checks if a field type occurs in its access path,
+    /// and adds diagnostics for recursive struct definitions if `ty` occurs in `path` and `ty` is also the struct type `checking`.
+    ///
+    /// Returns true iff no resursive definition for `checking` is found at this level.
+    ///
+    /// `path`: `path[0]` contains `path[1]` ... `path[-1]` contains `ty`
+    /// `checking`: the root struct which we are checking for recursive definition for
+    /// `loc_checking`: location of the struct we are checking for
+    fn check_recusive_struct_with_parents(
+        &self,
+        ty: &Type,
+        ty_loc: Loc,
+        path: &mut Vec<(Loc, Symbol, Type)>,
+        checking: QualifiedId<StructId>,
+        loc_checking: &Loc,
+    ) -> bool {
+        match ty {
+            Type::Struct(mid, sid, insts) => {
+                let this_struct_id = mid.qualified(*sid);
+                let this_struct_entry = self.lookup_struct_entry(this_struct_id);
+                // checks if `ty` occurs in `path`
+                for (_parent_loc, _field_symbol, parent) in path.iter() {
+                    if let Type::Struct(mid, sid, _) = parent {
+                        let parent_id = mid.qualified(*sid);
+                        if parent_id == this_struct_id {
+                            if checking == this_struct_id {
+                                let loop_notes = self.gen_error_msg_for_fields_loop(
+                                    path,
+                                    this_struct_id,
+                                    ty_loc.clone(),
+                                );
+                                self.error_with_notes(
+                                    loc_checking,
+                                    &format!(
+                                        "recursive definition {}",
+                                        self.get_struct_display_name_simple(this_struct_id)
+                                    ),
+                                    loop_notes,
+                                );
+                                return true;
+                            } else {
+                                return false;
+                            }
+                        }
+                    } else {
+                        panic!("field access path contains non-struct")
+                    }
+                }
+                if let Some(fields) = &this_struct_entry.fields {
+                    // check for descendant fields recursively
```

Comment:
If we consider the code:

```
struct X {
    a: Y,
    b: Y,
    c: Y,
    ..
    z: Y
}
```

where `Y` is a struct with several other fields (including possibly other structs, but none of them recursive), then we will explore a multiplicative number of paths (most of them unnecessarily)?

That is, if `X` has 10 fields that are `Y`s, and `Y` has 10 fields that are `Z`s (and `Z` is just a wrapper struct), then we recursively call this function 100 times, where as we only have to explore path `X` -any-field-> `Y` -any-field-> `Z` to show there is no recursive struct?

Reply:
solved in latest commit